torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,89 +65,142 @@ def reset_table(self) -> None:
-                self.generic_visit(node)
-                self.in_store = False
-                return
-            if node.func.attr in self.ALLOWED_READ_FNS:
-                self.read_depth += 1
-                self.generic_visit(node)
-                self.read_depth -= 1
-                return
-        self.generic_visit(node)
-
-
-def filter_non_mutated(kernel, tensors):
+    def __add__(self, o):
+        if isinstance(o, Proxy):
+            raise IdentifyMutationException("Proxy operation with Proxy")
+        return self
+
+
+class Scalar:
+    @staticmethod
+    def check(o):
+        if isinstance(o, Proxy):
+            raise IdentifyMutationException("Proxy operation with Scalar")
+
+    def __radd__(self, o):
+        if isinstance(o, Proxy):
+            return o
+        return self
+
+
+# This operation is only needed when triton is available
+if has_triton():
+    import triton
+
+    def replacement_fn(self, *args, **kwargs):
+        if len(args) > 0:
+            Scalar.check(args[0])
+        return self
+
+    for name, _ in inspect.getmembers(triton.language.core.tensor, inspect.isfunction):
+        if name in ["__init__", "__radd__"]:
+            continue
+
+        setattr(Scalar, name, replacement_fn)
+
+
+# Given a triton kernel and the arguments for this kernel, this function traces
+# through the triton kernel and identifies which input pointers are mutated.
+# Tracing is done by replacing the input pointers with Proxy objects that
+# track mutation. Each triton language function is monkey patched to
+# either detect the mutation or return a fresh scalar object.
+def identify_mutated_tensors(kernel, kwargs):
+    from triton import language as tl
     from triton.runtime.autotuner import Autotuner
 
+    # Replace tensor pointers with Proxy object. Do not replace the other
+    # inputs since tl.constexprs can be inspected by the kernel.
+    proxy_kwargs = {
+        key: (Proxy() if isinstance(value, Tensor) else value)
+        for key, value in kwargs.items()
+    }
+
     if isinstance(kernel, Autotuner):
+        if len(kernel.configs) > 0:
+            # If we are autotuning, then it doesn't matter which version gets
+            # picked for tracing purposes, so lets pick the first one
+            proxy_kwargs = {**proxy_kwargs, **kernel.configs[0].kwargs}
         kernel = kernel.fn
 
-    infos = {name: MutationInfo() for name in tensors}
-    tracker = MutationTracker(infos)
-    tracker.visit(kernel.parse())
-    return [
-        name for name, info in infos.items() if info.mutated or info.used_in_unknown
-    ]
+    default_impls: Dict[str, FunctionType] = {}
+    mutation_ops = {
+        "store",
+        "atomic_add",
+        "atomic_cas",
+        "atomic_max",
+        "atomic_min",
+        "atomic_xchg",
+        "make_block_ptr",
+    }
+    try:
+        # Monkey patch all triton language functions
+        for name, impl in inspect.getmembers(tl, inspect.isfunction):
```

Comment:
What will happen if the user does something like:
```python
from triton.language import store

@triton.jit
def foo(x):
   store(x, 0)
```

This is usually a pitfall of monkey patching since a copy of `store` is made in the user's scope.